common/predictive-text/virtualizedWorker.ts

```diff
@@ -0,0 +1,86 @@
+  // Unfortunately, importing like this loses all type information for the modules.
+  // It's that, or massively overhaul lm-layer to be module-based instead of
+  // namespace-based.
+  var fs = require("fs");
+  var vm = require("vm");
+
+  class VirtualizedWorkerContext {
+    // The LMLayerWorker installs itself to 'self', the expected Worker global, so we provide an alias.
+    self: VirtualizedWorkerContext;
+
+    constructor() {
+      this.self = this;
+    }
+
+    postMessage: (message: any, options: PostMessageOptions) => void;
+
+    importScripts() {
+      /* Use of vm.createContext and script.runInContext allow us to avoid
+        * polluting the global scope with imports.  When we throw away the
+        * context object, imported scripts will be automatically GC'd.
+        */
+      for(var i=0; i < arguments.length; i++) {
+        this.__importScriptString(fs.readFileSync(arguments[i]) as any as string);
+      }
+    }
+
+    __importScriptString(scriptStr: string) {
+      let context = vm.createContext(this);
+      var script = new vm.Script(scriptStr);
+      script.runInContext(context);
+    }
+  }
+
+  /**
+   * Note:  this does not create an actual Worker, separate process, or thread.  Everything will
+   *        be executed in-line on a virtualized context.
+   * 
+   *        In the future, it might be nice to use Node's Worker Threads implementation.
+   */
+  export class VirtualizedWorker implements Worker {
+    private _workerContext: VirtualizedWorkerContext;
+
+    constructor(scriptStr: string) {
+      this._workerContext = new VirtualizedWorkerContext();
+      // Needs to exist before setting up the worker; must exist by `.install()`.
+      this._workerContext.postMessage = this.workerPostMessage.bind(this);
+
+      // Initialize the "worker".
+      this._workerContext.__importScriptString(scriptStr);
+    }
+
+    // Sends the worker's postMessage messages to the appropriate `onmessage` handler.
+    private workerPostMessage(message: any, options: any) {
+      if(this.onmessage) {
+        this.onmessage({data: message} as any as MessageEvent);
+      }
+    }
+
+    onmessage: (this: Worker, ev: MessageEvent) => any;
+
+    postMessage(message: any, transfer: any[]): void;
+    postMessage(message: any, options?: PostMessageOptions): void;
+    postMessage(message: any, options?: any) {
+      let msgObj = {data: message};
+      let msgJSON = JSON.stringify(msgObj);
+
+      /* 
+       * Execute the command within the virtualized worker's scope.  The worker's returned 
+       * `postMessage` calls will still reach outside, as they have a reference to `this` via
+       * `postMessage` (which we've set to a bound `this.workerPostMessage`).
+       * 
+       * Among other things, this will allow the worker to use its internal namespaces without issue.
+       */
+      let msgCommand = "onmessage(" + msgJSON + ")";
+      this._workerContext.__importScriptString(msgCommand);
```

Comment:
The one concern I have here:  when the virtualized `Worker` calls its `postMessage`, that will call `this.workerPostMessage`, escaping the virtualized scope... but chances are that the worker's context will still be the `global` object at that time.

As long as we don't need to access that in callbacks, we should be fine.  In fact, it _is_ fine for now, with the LMLayer's current state.  Just noting a potential limitation of this approach.  (It's still better than losing access to the `Worker`'s `self` instance!  `Worker`s can usually afford to be far messier since they're self-contained.)

I did consider using a similar approach to reset to the original `global`, but there's the little detail that we'd need to call a method on the property of an instance - which would force us to (at least, temporarily) manipulate the global scope so that it'd be accessible.  In the worst case, that's a bridge we can cross if and when we come to it.